python/ray/dashboard/memory.py

```diff
@@ -0,0 +1,280 @@
+            "contained_in_owned": [
+                object_id.hex() for object_id in self.contained_in_owned
+            ],
+            "type": "Driver" if self.is_driver else "Worker"
+        }
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __repr__(self):
+        return str(self.__dict__())
+
+
+class MemoryTableSummary:
+    def __init__(self):
+        self.total_object_size = 0
+        self.total_local_ref_count = 0
+        self.total_pinned_in_memory = 0
+        self.total_used_by_pending_task = 0
+        self.total_captured_in_objects = 0
+        self.total_actor_handles = 0
+
+    def summarize(self, table: typing.List[MemoryTableEntry]):
+        for entry in table:
+            if entry.object_size > 0:
+                self.total_object_size += entry.object_size
+            if entry.reference_type == ReferenceType.LOCAL_REFERENCE:
+                self.total_local_ref_count += 1
+            elif entry.reference_type == ReferenceType.PINNED_IN_MEMORY:
+                self.total_pinned_in_memory += 1
+            elif entry.reference_type == ReferenceType.USED_BY_PENDING_TASK:
+                self.total_used_by_pending_task += 1
+            elif entry.reference_type == ReferenceType.CAPTURED_IN_OBJECT:
+                self.total_captured_in_objects += 1
+            elif entry.reference_type == ReferenceType.ACTOR_HANDLE:
+                self.total_actor_handles += 1
+
+    def __dict__(self):
+        return {
+            "total_object_size": self.total_object_size,
+            "total_local_ref_count": self.total_local_ref_count,
+            "total_pinned_in_memory": self.total_pinned_in_memory,
+            "total_used_by_pending_task": self.total_used_by_pending_task,
+            "total_captured_in_objects": self.total_captured_in_objects,
+            "total_actor_handles": self.total_actor_handles
+        }
+
+    def __str__(self):
+        return str(self.__dict__())
+
+
+class MemoryTable:
+    def __init__(self):
+        self.table: typing.List[MemoryTableEntry] = []
+        # Group is a list of memory tables grouped by a group key.
+        self.group: typing.Dict[str, MemoryTable] = self._get_resetted_group()
+        self.summary: MemoryTableSummary = MemoryTableSummary()
+
+    def insert_entry(self, entry: MemoryTableEntry):
+        """Insert a new memory table entry to the table."""
+        self.table.append(entry)
+
+    def setup(self,
+              *,
+              group_by_type: GroupByType = GroupByType.NODE_ADDRESS,
+              sort_by_type: SortingType = SortingType.PID):
+        """Setup memory table.
+
+        This will sort entries first and gruop them after.
+        Sort order will be still kept.
+        """
+        self.sort_by(sort_by_type).group_by(group_by_type)
+        for group_memory_table in self.group.values():
+            group_memory_table.summarize()
+        self.summarize()
+
+    def summarize(self):
```

Comment:
Correct me if I'm wrong, but I think the summary could also be calculated at initialization time so that it isn't possible for someone to forget to call summarize before reading the results of the table.